### PR TITLE
[Testcase Refactoring] Split test_checkpoint_wrapper by hardware classification

### PR DESCRIPTION
**Pull request by twthighd**
## Summary

This PR splits `test_checkpoint_wrapper.py` into a GENERIC pure-CPU class (`CheckpointWrapperTest`) and an ACCELERATOR device-dependent class (`CheckpointWrapperTestDevice`), labels both with `HardwareClassification`, and migrates the device-dependent tests to the `device` parameter injected by `instantiate_device_type_tests`. Review-feedback updates: `test_checkpoint_wrapper_parity` is un-skipped (its `@unittest.skip` is removed and it now runs and passes), its memory-stat calls use `torch.accelerator.*`, and `test_checkpoint_wrapper_with_block_mask` gets the standard `@onlyAccelerator` gate.

## Changes

- **`test/distributed/fsdp/test_checkpoint_wrapper.py`**:
  - Labeled `CheckpointWrapperTest` `hw_classification = HardwareClassification.GENERIC` (6 CPU-only checkpoint-wrapper logic methods: `test_load_activation_checkpointed_module`, `test_checkpoint_wrapper_kwarg_support`, `test_checkpoint_wrapper_args_kwargs`, `test_forward_missing_attributes`, `test_apply_activation_checkpointing`, `test_fqn` — all use unqualified CPU tensors, no accelerator dependency).
  - Extracted the 3 device-dependent methods (`test_checkpoint_wrapper_parity`, `test_checkpoint_wrapper_cpu_offload`, `test_checkpoint_wrapper_with_block_mask`) into a new `CheckpointWrapperTestDevice(TestCase)` labeled `hw_classification = HardwareClassification.ACCELERATOR`, instantiated via `instantiate_device_type_tests(..., except_for=("cpu",), allow_xpu=True)`.
  - Migrated all in-body device usage from the module-level `device_type = torch.device(get_devtype())` string to the injected `device` parameter: `.to(device_type.type)` → `.to(device)`, `randn(device=device_type.type)` → `randn(device=device)`, `torch.get_device_module(device_type.type)` → `torch.get_device_module(device)`. `CheckpointWrapperTestDevice` is a single-process `TestCase` (not `MultiProcessTestCase`), so bare indexed `device` is correct here (the type-only rule applies to multi-process per-rank classes only).
  - Decoupled `test_checkpoint_wrapper_with_block_mask` from hard-coded CUDA: `.cuda()` / `device="cuda"` use the injected `device`, and the old `@unittest.skipIf(not torch.cuda.is_available(), "requires CUDA")` is replaced by the standard method-level `@onlyAccelerator` gate (review feedback).
  - Removed the `@unittest.skip` on `test_checkpoint_wrapper_parity` so the test actually runs (review feedback) — verified passing on an 8-device accelerator backend; it was an unconditional, reason-less skip in the original file.
  - `test_checkpoint_wrapper_parity` now calls `torch.accelerator.reset_peak_memory_stats()` / `torch.accelerator.max_memory_allocated()` instead of `torch.get_device_module(device).<api>()` (review feedback).
  - Removed the now-unused module-level `device_type` / `get_devtype` import (and the `unittest` import once the skip decorators are gone).

## Motivation

An unclassified `TestCase` is silently dropped when `--hw-classification GENERIC ACCELERATOR` is passed. `CheckpointWrapperTest` was a mixed class: 6 CPU-only methods alongside 3 device-dependent methods. A single class cannot be labeled correctly when its methods span two classifications, so the device-dependent methods are split into `CheckpointWrapperTestDevice`. The `except_for=("cpu",)` mechanism replaces the hard-coded `only_for=("cuda","hpu","xpu")` whitelist so any accelerator registered via `torch.accelerator` (including PrivateUse1-based out-of-tree backends) is admitted. The ACCELERATOR class is structured to align with the TEST_LINTER direction (PR #190173): instantiated, `device` param on every `test_*`, no `only_for`.

## Test Plan

Verified on an 8-device accelerator backend (the `HardwareClassification` infrastructure from #186918 and the `skip_if_lt_x_gpu` device-agnostic rewrite from #187650 are pre-applied on the verify host):

```bash
python test/distributed/fsdp/test_checkpoint_wrapper.py
# Ran 9 tests in 1.238s — OK (9 passed; includes the now-un-skipped
# test_checkpoint_wrapper_parity_npu and the flex-attention BlockMask test
# test_checkpoint_wrapper_with_block_mask_npu, both passing on-device)

python test/distributed/fsdp/test_checkpoint_wrapper.py --hw-classification GENERIC ACCELERATOR
# total=9, passed=9, unclassified=0, mismatched=0
```

`CheckpointWrapperTestDevice` runs on-device via the injected `device` param; `CheckpointWrapperTest` (GENERIC) runs on CPU. `lintrunner` reports no lint issues.

## Notes

- `except_for=("cpu",)` is kept (not dropped) on `CheckpointWrapperTestDevice` because `test_checkpoint_wrapper_cpu_offload` has no per-method guard and tests "offload to CPU" — if the CPU variant were generated, it would pass vacuously (model already on CPU, offload is a no-op, `saved.device == cpu` is trivially true). `except_for=("cpu",)` excludes the CPU variant to keep that test meaningful. The `@onlyAccelerator` on `test_checkpoint_wrapper_with_block_mask` is the method-level equivalent of this gate (defensive/semantic — it keeps the accelerator requirement explicit at the method level).
- This is part of the test case refactoring initiative tracked in #185590.

### CODE REVIEW

**Comment by twthighd** [comment]
@wjlFlyer @pjyFight @FuDdd @fffrog @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

File: test/distributed/fsdp/test_checkpoint_wrapper.py [hunk 1]
```diff
@@ -16,18 +16,22 @@
...
 from torch.utils.checkpoint import checkpoint
 
 
 _SAVED_PREFIX = "_saved_"
 GRAD_FN_NEXT_FUNCTIONS = "next_functions"
 
-device_type = torch.device(get_devtype())
-
 
 class CheckpointWrapperTest(TestCase):
```
Comment (wjlFlyer) [inline test/distributed/fsdp/test_checkpoint_wrapper.py:34]: Class `CheckpointWrapperTest` isn't intantiated by `instantiate_device_type_tests`.

Comment (twthighd) [reply test/distributed/fsdp/test_checkpoint_wrapper.py:34]: Thanks for the review! Only ACCELERATOR-labeled classes need `instantiate_device_type_tests`. GENERIC classes (like `CheckpointWrapperTest`) are plain `TestCase` and don't require it.

**Review by wjlFlyer — APPROVED** [review]

**Review by wjlFlyer — APPROVED** [review]

File: test/distributed/fsdp/test_checkpoint_wrapper.py [hunk 2]
```diff
@@ -340,12 +265,95 @@ def test_fqn(self):
...
+
+        def test(use_checkpointing, use_wrapper, use_reentrant):
+            a = Model(
+                8,
+                use_checkpointing,
+                use_wrapper=use_wrapper,
+                use_reentrant=use_reentrant,
+            ).to(device)
+            x = torch.randn(10000, 256, requires_grad=True).to(device)
+            torch.get_device_module(device).reset_peak_memory_stats()
```
Comment (fffrog) [inline test/distributed/fsdp/test_checkpoint_wrapper.py]: Using torch.accelerator.reset_peak_memory_stats please.

File: test/distributed/fsdp/test_checkpoint_wrapper.py [hunk 3]
```diff
@@ -340,12 +265,95 @@ def test_fqn(self):
...
+            loss = a(x).sum()
+            loss.backward()
+            return torch.get_device_module(device).max_memory_allocated()
```
Comment (fffrog) [inline test/distributed/fsdp/test_checkpoint_wrapper.py]: Using torch.accelerator.max_memory_allocated please.

File: test/distributed/fsdp/test_checkpoint_wrapper.py [hunk 4]
```diff
@@ -391,8 +399,7 @@ def dfs(grad_fn):
 
         torch.autograd.graph.saved_tensors_hooks.__init__ = orig_init
 
-    @unittest.skipIf(not torch.cuda.is_available(), "requires CUDA")
-    def test_checkpoint_wrapper_with_block_mask(self):
```
Comment (fffrog) [inline test/distributed/fsdp/test_checkpoint_wrapper.py:395]: adding @onlyAccelerator here, please

File: test/distributed/fsdp/test_checkpoint_wrapper.py [hunk 5]
```diff
@@ -340,12 +265,95 @@ def test_fqn(self):
                 fqn in state_dict, msg=lambda msg: f"{msg}\n{fqn} not in state_dict."
             )
 
-    def test_checkpoint_wrapper_cpu_offload(self):
+
+class CheckpointWrapperTestDevice(TestCase):
+    hw_classification = HardwareClassification.ACCELERATOR
+
+    @unittest.skip
```
Comment (fffrog) [inline test/distributed/fsdp/test_checkpoint_wrapper.py]: Try to delete this one?

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/distributed/fsdp/test_checkpoint_wrapper.py (hunk 2 above)
Comment (twthighd) [reply test/distributed/fsdp/test_checkpoint_wrapper.py]: Thanks for the review! All four addressed:

@unittest.skip on test_checkpoint_wrapper_parity removed — the test now runs and passes on-device.
Its memory-stat calls switched to torch.accelerator.reset_peak_memory_stats() / torch.accelerator.max_memory_allocated().
test_checkpoint_wrapper_with_block_mask now has @onlyAccelerator.
Re-ran the full file: 9 tests OK (including the un-skipped parity test and the BlockMask test on-device).

**Review by fffrog — DISMISSED** [review]
SGTM.

File: test/distributed/fsdp/test_checkpoint_wrapper.py (hunk 3 above)
Comment (twthighd) [reply test/distributed/fsdp/test_checkpoint_wrapper.py]: Done — switched to torch.accelerator.max_memory_allocated(); see my reply in the first comment thread for the full summary.

File: test/distributed/fsdp/test_checkpoint_wrapper.py (hunk 4 above)
Comment (twthighd) [reply test/distributed/fsdp/test_checkpoint_wrapper.py:395]: Done — @onlyAccelerator added to test_checkpoint_wrapper_with_block_mask; see my reply in the first comment thread.

File: test/distributed/fsdp/test_checkpoint_wrapper.py (hunk 5 above)
Comment (twthighd) [reply test/distributed/fsdp/test_checkpoint_wrapper.py]: Done — removed the @unittest.skip; the test now runs and passes on-device. See my reply in the first comment thread.

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

**Comment by twthighd** [comment]
> @claude review this pr with pr review skill

Thanks for the thorough review — all actionable findings are addressed in bb00b0c5:

create_block_mask now receives device=device (the mask is built on the parametrized device instead of the current accelerator).
The parity memory-stat calls now pass the injected device explicitly: torch.accelerator.reset_peak_memory_stats(device) / max_memory_allocated(device) — the no-arg forms read the current device index, which desyncs from the parametrized variant after any set_device.
Non-zero sanity assertions added before both parity equality checks, so an uninitialized allocator (which reports peak 0) now fails loudly instead of passing vacuously.
Gating unified: all three CheckpointWrapperTestDevice methods carry @onlyAccelerator and except_for=("cpu",) is dropped from the instantiate call — the TEST_LINTER-aligned shape; CPU-only hosts now report 3 explicit skips instead of 0 collected. This also supersedes the old Notes rationale for except_for: the vacuous-pass concern for cpu_offload on the CPU variant is covered by the method-level gate.

On the un-skip: it was reviewer-directed in this PR (fffrog asked for the skip removal) and verified on-device (8-card out-of-tree accelerator, exact peak-memory equality passes); the PR-triggered workflows have no GPU runners, so the CUDA run will be the nightly multigpu job after merge. The original skip was unconditional with no reason string, so its rationale is not recoverable
